### ISSUE TITLE
Wrong colors/color skewing when using global seam leveling

### ISSUE THREAD

**Issue by svetlins** [nmoehrle/mvs-texturing#63]
Hello,

I'm trying to texture a mesh with some aerial imagery (as part of a larger toolchain - [OpenDroneMap](https://github.com/OpenDroneMap/OpenDroneMap)). The results I'm getting are somewhat wrong and I've traced the color skewing to mvs-texturing and more specifically to the the global seam leveling option. [Here](https://www.dropbox.com/s/wtkno07pt58ckhs/example.zip?dl=0) is the nvm, images and ply file as well as screen shots of the results I'm comparing.

Here is the result with `--skip_global_seam_leveling`:
<img width="319" alt="good" src="https://cloud.githubusercontent.com/assets/90983/21311001/b61467fa-c5ed-11e6-9eeb-0e2a1adadaef.png">
(exact command used: `texrecon pmvs/nvmCams.nvm mesh.ply output/good -d gmi -o none --skip_global_seam_leveling`)

And here is without:
<img width="326" alt="bad" src="https://cloud.githubusercontent.com/assets/90983/21311421/cbbbc2b8-c5ef-11e6-9e05-0b46cbdc05f7.png">
(exact command used: `texrecon pmvs/nvmCams.nvm mesh.ply output/bad -d gmi -o none`)

The texrecon executable was built by me against 93a261f1dd42e47b9c20d5ef44d9663fc0fa3233 which was the latest master at the time of writing this issue.

One interesting tidbit/observation that may as well be wrong but it looks like this color skewing is noticeable on all aerial datasets I've tested with but is more pronounced on ones that were shot with some kind of color filter as this one is. 

I'm ready to provide any other details if you need them.

And thanks a lot for your work!

**Comment by rionlerm** [comment]
Another Case study:

[https://cloud.githubusercontent.com/assets/13643064/21981199/ee7c846c-dbee-11e6-8ccd-2a74cb632c5f.png](url)

**Comment by rionlerm** [comment]
Mesh and Orthophoto anomalies issue #454

![ee7c846c-dbee-11e6-8ccd-2a74cb632c5f](https://cloud.githubusercontent.com/assets/13643064/22024183/c7b45164-dcd1-11e6-8d8a-27525f942f5f.png)

**Comment by nmoehrle** [comment]
Thanks, for raising the issue and providing an example.
I am sorry that I haven't addressed this issue yet and that I cannot tell you when I will have time to do so...

**Comment by dakotabenjamin** [comment]
I will also provide another example of color balance issues. Maybe it is related to #59 as well?

![odm_orthophoto-scaled](https://cloud.githubusercontent.com/assets/1847818/22026843/f49e62a0-dc9f-11e6-8de5-57856fa0b201.png)

**Comment by smathermather** [comment]
Hi @nmoehrle -- is there any way we can help, either by providing development resources or other issues?

**Comment by nmoehrle** [comment]
Thanks for the offer!

I have two deadlines coming up that currently require all of my attention so I cannot experiment with these datasets myself.

That being said I could not help myself and had a look at your first dataset. If you reconstruct it without any seam leveling you see how different the colors of the input image actually are. The issue here is that our global seam optimization has due to the view selection little to no constraints on how to adjust the images. There are many regions where it has only three image, with wildly different colors, that separate all other images so it simply distributes the errors.

![no-seam-leveling](https://cloud.githubusercontent.com/assets/4200007/22030978/937ee4fa-dcdf-11e6-85d7-0c9b3ef337f2.png)
![view-selection](https://cloud.githubusercontent.com/assets/4200007/22030979/9384f19c-dcdf-11e6-85f9-d8e7dcd7bc4a.png)

The only Idea that I currently have is regularizing the global adjustments. In datasets with less boundaries this happens implicitly and a regularization would hurt the quality there but for these datasets it might be necessary.

The relevant section in the [paper](http://www.gcc.tu-darmstadt.de/media/gcc/papers/Waechter-2014-LTB.pdf) is 4.3 equation (3) the implementation is in [global_seam_leveling.cpp#L243](https://github.com/nmoehrle/mvs-texturing/blob/master/libs/tex/global_seam_leveling.cpp#L243). In theory it should be possible to simply multiply with another regularization matrix that penalizes adjustments in general. If you search for a reference the [wiki](https://en.wikipedia.org/wiki/Tikhonov_regularization) entry for Tikhonov regularizations is actually quite good.

The issues of the other datasets look different to me, I belief that alignment (and thus the adjustment of non corresponding colors) is more of an issue there.

I am happy to answer questions but I cannot dig into it myself right now.

Best,
Nils

**Comment by nmoehrle** [comment]
Only global seam leveling without regularization:
![before](https://cloud.githubusercontent.com/assets/4200007/22081022/27e9d7de-ddc2-11e6-8c95-69927cac1055.png)
Only global seam leveling with regularization:
![after](https://cloud.githubusercontent.com/assets/4200007/22081023/291343e8-ddc2-11e6-9ead-17ae4074b877.png)

Seams like a regularization could work - could you guys do some tuning and testing?

**Comment by dakotabenjamin** [comment]
I have a few datasets run with a lens filter so yes I can do that.

**Comment by svetlins** [comment]
@nmoehrle that original dataset looks much better! I will run a few other datasets and will post results here. (I assume I need to build the `regularization` branch)

Thanks again for your awesome work! :)

**Comment by nmoehrle** [comment]
Yes the regularization is the multiplication with a weighted identity matrix in this [line](https://github.com/nmoehrle/mvs-texturing/commit/e695a29505c74be570b8bbb87cf42f5b1611cf60#diff-9c0b2a2a8a92d76aa42180ddf18fdde9R246) that puts a quadratic penalty on all adjustments. The hard coded weight is the first thing that came to my mind - so its arbitrary - and requires tuning or should be exposed to the command line...

**Comment by dakotabenjamin** [comment]
I will be testing on this dataset: https://github.com/OpenDroneMap/odm_data_seneca You can see the mosaic I've generated from the original mvstex results in the README.

**Comment by dakotabenjamin** [comment]
For tuning, what kind of range are we looking at? This kind of math is far beyond my understanding, so the best I can do is guess and check the numbers. 

So far the results look good for me using the default 0.0001
![image](https://cloud.githubusercontent.com/assets/1847818/22114577/054da7f2-de38-11e6-8265-a45bb5b29428.png)

**Comment by dakotabenjamin** [comment]
Tuning orders of magnitude, 0.0001, 0.001, 0.00001
https://clemetparks-my.sharepoint.com/personal/dmb2_clevelandmetroparks_com/_layouts/15/guestaccess.aspx?docid=156a2a039a8b14d14b1005111e488d4cc&authkey=ATO2sgKPA0N0qvVueStyiLo

Sorry for the bad file labeling.

**Comment by dakotabenjamin** [comment]
Visually, 0.0001 looks better. 0.00001 has too much darkening along the edges and 0.001 has these patches of light and dark. I would say the former does the best job accounting for some of the vignetting and cloud-cover issues from the original dataset.

**Comment by nmoehrle** [comment]
Wow, lucky guess I guess ^^

Indeed with 1e-3 we can already see how the leveling is degrading and 1e-5 is not strong enough for the borders, but I think this will depend somewhat on the dataset so I wont merge it into master right away at least not without exposing it to the command line.

It would be interesting to see the results when we only regularize the borders...

Thanks for testing!

**Comment by smathermather** [comment]
Looks promising:
https://github.com/OpenDroneMap/odm_data_seneca/commit/fec053a3c77eb80a2a445b3a03d50af5eaa77e5f

https://github.com/smathermather/OpenDroneMap/tree/mvs-global-illumination-patch

**Comment by rionlerm** [comment]
Definitely,

gonna configure ODM and test on some data.

Regards,
*Rion Lerm (jnr.)*
(+27) 76 913 8381 (Cell)
(+27) 13 735 3536 (Office)
Ndlovu Node
Phalaborwa Gate, Kruger National Park
South African Environmental Observation Network (SAEON)

Click here to learn about SAEON <http://www.saeon.ac.za/saeon-background>

**Comment by rionlerm** [comment]
Before:
![odm_orthophoto_resized](https://cloud.githubusercontent.com/assets/13643064/22732641/9006f9c4-edf7-11e6-9ae3-bbce16052e6e.png)

After:
![odm_orthophoto_mvs-texturing_resized](https://cloud.githubusercontent.com/assets/13643064/22732665/a88c4260-edf7-11e6-8e48-ffa7a5214a28.png)

**Comment by nmoehrle** [comment]
Was there a cloud? Or is it still failing?

**Comment by rionlerm** [comment]
Cloud.

**Comment by smathermather** [comment]
Outstanding. Dakota and I are testing a couple of other tricky datasets today. We'll let you know how it goes... .
